spices/SPICE-0004-HTTP-Proxy-Support.adoc

```diff
@@ -0,0 +1,180 @@
+/// ```
+address: Uri(startsWith("http://"))
+
+/// Hosts to disable connecting to via a proxy.
+///
+/// Values can be either hostnames, or CIDR blocks.
+///
+/// Follows the same rules as cURL's [--noproxy](https://curl.se/docs/manpage.html#--noproxy)
+/// flag.
+noProxy: Listing<String>(isDistinct)
+----
+
+==== CLI flags
+
+The following CLI flags are added to common CLI options:
+
+* `--proxy` -- A proxy URL to connect to
+* `--noproxy` -- A comma-separated list of hosts to not proxy.
+
+==== pkl-gradle
+
+The following options are added to the common options for `PklSpec`:
+
+* `proxy: Property<String>`
+* `noProxy: ListProperty<String>`
+
+==== `HttpClient#Builder`
+
+When Pkl is embedded with in Java, it is preferrable to use the host's proxy settings if possible.
+When Pkl is run standalone, the proxy should be configued via user-facing means.
+
+To support this, `HttpClient#Builder` receives two new builder methods:
+
+* `setProxySelector(java.net.ProxySelector proxySelector)`
+* `setProxy(URI proxyAddress, List<URI> noProxy)`
+
+The default builder receives `java.net.ProxySelector#getDefault`, which uses host proxy settings if one is set.
+This is the proxy selector that gets used by `EvaluatorBuilder#preconfigured`.
+
+`setProxy(URI proxyAddress, List<URI> noProxy)` sets a proxy selector that follows the noProxy rules described in <<noproxy>>,
+and is what the CLI, Gradle, and message passing API use.
+
+=== Proxy Configuration Values
+
+There are two values to configure for proxying:
+
+==== Address
+
+The proxy address specifies the proxy host, protocol, and optional port.
+It takes the form of a standard URL. For example:
+
+[source,text]
+----
+"http://my.proxy.com:8080"
+----
+
+The scheme _must_ be `http`.
+
+[[noproxy]]
+==== `noProxy`
+
+The noProxy value follows the specification outlined in link:https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/#standardizing-no_proxy[standardizing `noproxy`], except in how it is configured.
+
+1. The `*` wildcard matches all hosts.
+2. A hostname with a leading dot (`.`) is stripped.
+3. A hostname matches all domain prefixes.
+4. CIDR blocks can be specified for matching.
+5. Hostnames are not resolved; a hostname will match its declared name only and not its possibly resolved IP address.
+6. IP addresses are treated verbatim (no special handling of loopback address).
+
+== Compatibility
+
+Java applications that embed Pkl and construct an evaluator with `EvaluatorBuilder#preconfigured` will be configured to talk to the host's proxy settings, if one is set via `java.net.ProxySelector.setDefault()`.
+This is technically a breaking change.
+
+== Future directions
+
+There are several more features related to proxies that can be supported in the future:
+
+* HTTPS proxies
+* SOCKS proxies
+* Alternative authentication methods (e.g. setting the `Proxy-Authorization` header)
+
+== Alternatives considered
+
+TBD
```

Comment:
URL rewriting was a considered alternative, but after some thought we found out they don't fully overlap and there may be a need for both.